### PR TITLE
Replace MultiClusterService with ServiceImport

### PR DESCRIPTION
**Pull request by aswinsuryan**
The ServiceImport replaces the MultiClusterService.

Signed-off-by: Aswin Surayanarayanan <asuryana@redhat.com>

### CODE REVIEW

File: pkg/agent/controller/agent.go [hunk 1]
```diff
@@ -78,23 +78,31 @@ func (a *Controller) Run(stopCh <-chan struct{}) error {
...
 			Annotations: map[string]string{
 				"origin-name":      svc.Name,
 				"origin-namespace": svc.Namespace,
 			},
 		},
-		Spec: lighthousev1.MultiClusterServiceSpec{
-			Items: []lighthousev1.ClusterServiceInfo{
+		Spec: lighthousev2a1.ServiceImportSpec{
+			Ports:                 nil,
+			IP:                    "10.10.10.10",
```
Comment (vthapar) [inline pkg/agent/controller/agent.go:99]: Is this because this is a required field? This should be optional since it will be empty for Headless service.

File: pkg/agent/controller/agent.go [hunk 2]
```diff
@@ -78,23 +78,31 @@ func (a *Controller) Run(stopCh <-chan struct{}) error {
...
+			Type:                  "",
```
Comment (vthapar) [inline pkg/agent/controller/agent.go:100]: Why are we not filling type?

File: pkg/agent/controller/agent.go [hunk 3]
```diff
@@ -136,14 +144,14 @@ func (a *Controller) updateExportedServiceStatus(export *lighthousev2a1.ServiceE
 	return retryErr
 }
 
-func (a *Controller) newClusterServiceInfo(service *corev1.Service, clusterID string) lighthousev1.ClusterServiceInfo {
+func (a *Controller) newClusterServiceInfo(service *corev1.Service, clusterID string) lighthousev2a1.ClusterStatus {
```
Comment (vthapar) [inline pkg/agent/controller/agent.go:147]: Rename method name to reflect the new object it returns

File: pkg/agent/controller/agent.go [hunk 4]
```diff
@@ -78,23 +78,31 @@ func (a *Controller) Run(stopCh <-chan struct{}) error {
...
 }
 
-func (a *Controller) serviceExportToRemoteMcs(obj runtime.Object) runtime.Object {
+func (a *Controller) serviceExportToRemoteMcs(obj runtime.Object) (runtime.Object, bool) {
 	svcExport := obj.(*lighthousev2a1.ServiceExport)
 	svc, err := a.kubeClientSet.CoreV1().Services(svcExport.Namespace).Get(svcExport.Name, metav1.GetOptions{})
 	if err != nil {
 		klog.Errorf("No matching service for %v", svcExport)
-		return nil
+		return nil, true
```
Comment (vthapar) [inline pkg/agent/controller/agent.go:86]: Need to handle delete use case here. Should wait till we get submariner-io/admiral#83 merged.

File: pkg/agent/controller/agent.go (hunk 1 above)
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go:99]: oh it was a code added to debug,, will remove it.

File: pkg/agent/controller/agent.go (hunk 2 above)
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go]: filled it with superclusterIP

File: pkg/agent/controller/agent.go (hunk 3 above)
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go]: Done.

File: scripts/build-agent [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Comment (vthapar) [inline scripts/build-agent:1]: Why are all these changes in this patch? Rebase issue?

File: pkg/serviceimport/map.go [hunk 6]
```diff
@@ -1,9 +1,9 @@
-package multiclusterservice
+package serviceimport
 
 import (
 	"sync"
 
-	lighthousev1 "github.com/submariner-io/lighthouse/pkg/apis/lighthouse.submariner.io/v1"
+	lighthousev2a1 "github.com/submariner-io/lighthouse/pkg/apis/lighthouse.submariner.io/v2alpha1"
```
Comment (vthapar) [inline pkg/serviceimport/map.go:6]: I'd like to see this moved from plugin to agent. The agent should aggregate ServiceImports from multiple clusters into single object and plugin uses it. It will allow us to be compatible with u/s API and won't need changes when it becomes available in K8s.

Comment (aswinsuryan) [reply pkg/serviceimport/map.go:6]: I can push it as another patch with those changes.

File: scripts/build-agent (hunk 5 above)
Comment (aswinsuryan) [reply scripts/build-agent]: yes , I removed it.

File: pkg/agent/controller/agent.go (hunk 4 above)
Comment (tpantelis) [reply pkg/agent/controller/agent.go:86]: It is merged but I think we should leave such changes for https://github.com/submariner-io/lighthouse/pull/188.  The purpose of this PR is only to switch to using `ServiceImport`. In fact I don't this PR should bump the admiral dependency to pick up the transform function signature changes - those are intended for  https://github.com/submariner-io/lighthouse/pull/188.

File: pkg/agent/controller/agent.go [hunk 7]
```diff
@@ -78,24 +78,32 @@ func (a *Controller) Run(stopCh <-chan struct{}) error {
...
 		},
-		Spec: lighthousev1.MultiClusterServiceSpec{
-			Items: []lighthousev1.ClusterServiceInfo{
-				a.newClusterServiceInfo(svc, a.clusterID),
+		Spec: lighthousev2a1.ServiceImportSpec{
+			Ports:                 nil,
+			IP:                    "",
+			Type:                  lighthousev2a1.SuperclusterIP,
+			SessionAffinity:       "",
+			SessionAffinityConfig: nil,
```
Comment (tpantelis) [inline pkg/agent/controller/agent.go:102]: It's not necessary to explicitly initialize these fields too their nil values as that's done automatically.

File: pkg/serviceimport/controller_test.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ = Describe("ServiceImport controller", func() {
+	klog.InitFlags(nil)
+
+	Describe("ServiceImport lifecycle notifications", testMCSLifecycleNotifications)
+})
+
+func testMCSLifecycleNotifications() {
```
Comment (tpantelis) [inline pkg/serviceimport/controller_test.go:24]:
```suggestion
func testLifecycleNotifications() {
```

File: pkg/serviceimport/suite_test.go [hunk 9]
```diff
@@ -1,4 +1,4 @@
-package multiclusterservice
+package serviceimport
```
Comment (tpantelis) [inline pkg/serviceimport/suite_test.go:1]: The UI won't let me comment on the lines below, so...

L10: rename to `TestServiceImport`
L12: "_MultiClusterService_" -> "_ServiceImport_"

File: plugin/lighthouse/handler_test.go [hunk 10]
```diff
@@ -227,14 +226,14 @@ func executeTestCase(lh *Lighthouse, rec *dnstest.Recorder, tc test.Case) {
 	}
 }
 
-func setupMultiClusterServiceMap() *multiclusterservice.Map {
-	mcsMap := multiclusterservice.NewMap()
-	mcsMap.Put(newMultiClusterService(namespace1, service1, serviceIP, "clusterID"))
+func setupServiceImportMap() *serviceimport.Map {
+	mcsMap := serviceimport.NewMap()
```
Comment (tpantelis) [inline plugin/lighthouse/handler_test.go:230]:
```suggestion
	siMap := serviceimport.NewMap()
```

File: plugin/lighthouse/handler_test.go [hunk 11]
```diff
@@ -227,14 +226,14 @@ func executeTestCase(lh *Lighthouse, rec *dnstest.Recorder, tc test.Case) {
...
+	mcsMap.Put(serviceImport(namespace1, service1, serviceIP, "clusterID"))
 	return mcsMap
 }
 
-func newMultiClusterService(namespace, name, serviceIP, clusterID string) *v1.MultiClusterService {
-	return &lighthousev1.MultiClusterService{
+func serviceImport(namespace, name, serviceIP, clusterID string) *lighthousev2a1.ServiceImport {
```
Comment (tpantelis) [inline plugin/lighthouse/handler_test.go:235]:
```suggestion
func newServiceImport(namespace, name, serviceIP, clusterID string) *lighthousev2a1.ServiceImport {
```

File: plugin/lighthouse/setup.go [hunk 12]
```diff
@@ -52,8 +52,8 @@ func lighthouseParse(c *caddy.Controller) (*Lighthouse, error) {
 		return nil, fmt.Errorf("error building kubeconfig: %v", err)
 	}
 
-	mcsMap := multiclusterservice.NewMap()
-	mcsController := multiclusterservice.NewController(mcsMap)
+	mcsMap := serviceimport.NewMap()
```
Comment (tpantelis) [inline plugin/lighthouse/setup.go:55]:
```suggestion
	siMap := serviceimport.NewMap()
```

File: plugin/lighthouse/setup.go [hunk 13]
```diff
@@ -52,8 +52,8 @@ func lighthouseParse(c *caddy.Controller) (*Lighthouse, error) {
...
+	mcsController := serviceimport.NewController(mcsMap)
```
Comment (tpantelis) [inline plugin/lighthouse/setup.go:56]:
```suggestion
	siController := serviceimport.NewController(mcsMap)
```

File: pkg/agent/controller/agent.go (hunk 4 above)
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go]: Removed the changes.

File: pkg/agent/controller/agent.go (hunk 7 above)
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go]: Done.

File: pkg/serviceimport/controller_test.go (hunk 8 above)
Comment (aswinsuryan) [reply pkg/serviceimport/controller_test.go]: Done.

File: pkg/serviceimport/suite_test.go (hunk 9 above)
Comment (aswinsuryan) [reply pkg/serviceimport/suite_test.go:1]: Done.

File: plugin/lighthouse/handler_test.go (hunk 10 above)
Comment (aswinsuryan) [reply plugin/lighthouse/handler_test.go]: Done.

File: plugin/lighthouse/handler_test.go (hunk 11 above)
Comment (aswinsuryan) [reply plugin/lighthouse/handler_test.go]: Done.

File: plugin/lighthouse/setup.go (hunk 12 above)
Comment (aswinsuryan) [reply plugin/lighthouse/setup.go]: Done.

File: plugin/lighthouse/setup.go (hunk 13 above)
Comment (aswinsuryan) [reply plugin/lighthouse/setup.go]: Done.

File: pkg/serviceimport/suite_test.go [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestServiceImport(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "ServiceImport	 Suite")
```
Comment (tpantelis) [inline pkg/serviceimport/suite_test.go:12]:
```suggestion
	RunSpecs(t, "ServiceImport Suite")
```

File: pkg/agent/controller/agent.go [hunk 15]
```diff
@@ -3,18 +3,18 @@ package controller
 import (
 	"fmt"
 
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/util/retry"
```
Comment (tpantelis) [inline pkg/agent/controller/agent.go:7]: Minor thing but why were these moved up here separated from the other `k8s.io` imports?

File: Dockerfile.dapper707020469 [hunk 16]
```diff
@@ -0,0 +1,10 @@
+FROM quay.io/submariner/shipyard-dapper-base:devel
```
Comment (vthapar) [inline Dockerfile.dapper707020469:1]: This file should be removed.

File: pkg/agent/controller/agent.go [hunk 17]
```diff
@@ -78,24 +77,28 @@ func (a *Controller) Run(stopCh <-chan struct{}) error {
...
 }
 
-func (a *Controller) serviceExportToRemoteMcs(obj runtime.Object) runtime.Object {
+func (a *Controller) serviceExportToRemoteServiceImport(obj runtime.Object) (runtime.Object, bool) {
 	svcExport := obj.(*lighthousev2a1.ServiceExport)
 	svc, err := a.kubeClientSet.CoreV1().Services(svcExport.Namespace).Get(svcExport.Name, metav1.GetOptions{})
 	if err != nil {
 		klog.Errorf("No matching service for %v", svcExport)
-		return nil
+		return nil, true
```
Comment (vthapar) [inline pkg/agent/controller/agent.go:85]: True will result in requeue. Set to false to retain existing behavior. As Tom said, this will be handled by #188 or #195

Comment (aswinsuryan) [reply pkg/agent/controller/agent.go:85]: I added this back since we Status will be distributed only in latest admiral. I will make it false.

File: Dockerfile.dapper707020469 (hunk 16 above)
Comment (aswinsuryan) [reply Dockerfile.dapper707020469]: Removed, I think we need gitignore for this .

File: pkg/agent/controller/agent.go (hunk 15 above)
Comment (aswinsuryan) [reply pkg/agent/controller/agent.go]: Done.

File: pkg/serviceimport/suite_test.go (hunk 14 above)
Comment (aswinsuryan) [reply pkg/serviceimport/suite_test.go:12]: Done.

File: go.mod [hunk 18]
```diff
@@ -10,7 +10,7 @@ require (
 	github.com/onsi/ginkgo v1.14.0
 	github.com/onsi/gomega v1.10.1
 	github.com/pkg/errors v0.9.1
-	github.com/submariner-io/admiral v0.4.1
+	github.com/submariner-io/admiral v0.4.2-0.20200706185424-d52c64ccd7ba
```
Comment (tpantelis) [inline go.mod:13]: You had removed the admiral upgrade before - did you mean to add it back or did it sneak back in?

File: pkg/agent/controller/agent.go (hunk 17 above)
Comment (tpantelis) [reply pkg/agent/controller/agent.go]: Ahh - ok - I noticed you had added it back and was going to comment...
